MFVNeutralino/plugins/Vertexer.cc

```diff
@@ -1,891 +1,2008 @@
+
+      h_output_aftersharedjets_n_onetracks = fs->make<TH1F>("h_output_aftersharedjets_n_onetracks", "", 5, 0, 5);
+    }
+  }
+}
+
+void MFVVertexer::finish(edm::Event& event, const std::vector<reco::TransientTrack>& seed_tracks, std::unique_ptr<reco::VertexCollection> vertices, std::unique_ptr<VertexerPairEffs> vpeffs, const std::vector<std::pair<track_set, track_set>>& vpeffs_tracks) {
+  std::unique_ptr<reco::TrackCollection> tracks_seed      (new reco::TrackCollection);
+  std::unique_ptr<reco::TrackCollection> tracks_inVertices(new reco::TrackCollection);
+
+  if (verbose) printf("finish:\nseed tracks:\n");
+
+  std::map<std::pair<unsigned, unsigned>, unsigned char> seed_track_ref_map;
+  unsigned char itk = 0;
+  for (const reco::TransientTrack& ttk : seed_tracks) {
+    tracks_seed->push_back(ttk.track());
+    const reco::TrackBaseRef& tk(ttk.trackBaseRef());
+    seed_track_ref_map[std::make_pair(tk.id().id(), tk.key())] = uint2uchar_clamp(itk++);
+
+    if (verbose) printf("id: %i key: %lu pt: %f\n", tk.id().id(), tk.key(), tk->pt());
+  }
+
+  assert(vpeffs->size() == vpeffs_tracks.size());
+  for (size_t i = 0, ie = vpeffs->size(); i < ie; ++i) {
+    for (auto tk : vpeffs_tracks[i].first)  (*vpeffs)[i].tracks_push_back(0, seed_track_ref_map[std::make_pair(tk.id().id(), tk.key())]);
+    for (auto tk : vpeffs_tracks[i].second) (*vpeffs)[i].tracks_push_back(1, seed_track_ref_map[std::make_pair(tk.id().id(), tk.key())]);
+  }
+
+  if (verbose) printf("vertices:\n");
+  int count_5trk_vertices = 0;
+  for (const reco::Vertex& v : *vertices) {
+    if (verbose) printf("x: %f y %f z %f\n", v.x(), v.y(), v.z());
+    for (auto it = v.tracks_begin(), ite = v.tracks_end(); it != ite; ++it) {
+      reco::TrackRef tk = it->castTo<reco::TrackRef>();
+      if (verbose) printf("id: %i key: %u <%f,%f,%f,%f,%f>\n", tk.id().id(), tk.key(), tk->charge()*tk->pt(), tk->eta(), tk->phi(), tk->dxy(), tk->dz());
+      tracks_inVertices->push_back(*tk);
+    }
+    if (v.nTracks() >= 5)
+      ++count_5trk_vertices;
+  }
+
+  if (verbose)
+    printf("n_output_vertices: %lu\n", vertices->size());
+  if (histos) {
+    h_n_output_vertices->Fill(vertices->size());
+    h_n_at_least_5trk_output_vertices->Fill(count_5trk_vertices);
+  }
+
+  event.put(std::move(vertices));
+  event.put(std::move(vpeffs));
+  event.put(std::move(tracks_seed),       "seed");
+  event.put(std::move(tracks_inVertices), "inVertices");
+}
+
+void MFVVertexer::produce(edm::Event& event, const edm::EventSetup& setup) {
+  if (verbose)
+    std::cout << "MFVVertexer " << module_label << " run " << event.id().run() << " lumi " << event.luminosityBlock() << " event " << event.id().event() << "\n";
+
+  edm::Handle<reco::BeamSpot> beamspot;
+  event.getByToken(beamspot_token, beamspot);
+  const double bsx = beamspot->position().x();
+  const double bsy = beamspot->position().y();
+  const double bsz = beamspot->position().z();
+

```

Comment:
I think it would be reasonable to put
`    const reco::Vertex fake_bs_vtx(beamspot->position(), beamspot->covariance3D());`

here, and to drop it from all of its later instances, so that we don't need to create this object 5x.

Reply:
Is the one you're working on having a function of making all identical plots too?

Reply:
Yes! Done in https://github.com/DisplacedVertices/cmssw-usercode/commit/c7e18622d9dd7ed5b51ea31fa16ffab25e709090. Double check my work though!

Reply:
The test run works now and all plots by hist types have entries filled. It was just the mass/vtx that did not have entries filled (and that was a error propagation from my code before your simplification). All good now.